Failure to delete pending snapshot leads to inactive partition until restart

**Describe the bug**

If a snapshot is being replicated during startup, and it fails to delete an existing pending snapshot directory (could be a leftover from a crash, or simply the replication restarting due to some errors), the partition transitions to inactive and is stuck, preventing startup from finishing.

**To Reproduce**

It seems to happen more often with large state, >= 8GB. Simply cause data loss on one node to force replication on restart, and it might happen.

**Expected behavior**

Even if an I/O error occurs during replication, it should be able to recover and retry the snapshot replication.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
java.lang.IllegalStateException: Expected to delete pending received snapshot, but failed.
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.newReceivedSnapshot(FileBasedSnapshotStore.java:373) ~[zeebe-snapshots-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.newReceivedSnapshot(FileBasedSnapshotStore.java:51) ~[zeebe-snapshots-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.raft.roles.PassiveRole.onInstall(PassiveRole.java:191) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.raft.roles.FollowerRole.onInstall(FollowerRole.java:83) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.raft.impl.RaftContext.lambda$registerHandlers$8(RaftContext.java:315) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.raft.impl.RaftContext.lambda$runOnContext$27(RaftContext.java:334) ~[zeebe-atomix-cluster-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:178) [zeebe-atomix-utils-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.nio.file.NoSuchFileException: /usr/local/zeebe/data/raft-partition/partitions/2/snapshots/6651-18-6637-6635/001698.sst
	at sun.nio.fs.UnixException.translateToIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.implDelete(Unknown Source) ~[?:?]
	at sun.nio.fs.AbstractFileSystemProvider.delete(Unknown Source) ~[?:?]
	at java.nio.file.Files.delete(Unknown Source) ~[?:?]
	at io.camunda.zeebe.util.FileUtil$FolderDeleter.visitFile(FileUtil.java:204) ~[zeebe-util-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.camunda.zeebe.util.FileUtil$FolderDeleter.visitFile(FileUtil.java:187) ~[zeebe-util-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at java.nio.file.Files.walkFileTree(Unknown Source) ~[?:?]
	at java.nio.file.Files.walkFileTree(Unknown Source) ~[?:?]
	at io.camunda.zeebe.util.FileUtil.deleteFolder(FileUtil.java:134) ~[zeebe-util-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	at io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore.newReceivedSnapshot(FileBasedSnapshotStore.java:371) ~[zeebe-snapshots-8.4.0-SNAPSHOT.jar:8.4.0-SNAPSHOT]
	... 12 more
```

</p>
</details>

**Environment:**
- Zeebe Version: 8.4.0-SNAPSHOT

--- a/atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java
+++ b/atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java
@@ -53,6 +53,7 @@ import io.camunda.zeebe.util.logging.ThrottledLogger;
 import java.nio.ByteBuffer;
 import java.time.Duration;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
 import org.agrona.concurrent.UnsafeBuffer;
 
 /** Passive state. */
@@ -193,14 +194,24 @@ public class PassiveRole extends InactiveRole {
 
       try {
         pendingSnapshot =
-            raft.getPersistedSnapshotStore().newReceivedSnapshot(snapshotChunk.getSnapshotId());
-      } catch (final SnapshotAlreadyExistsException snapshotAlreadyExists) {
-        // This should not happen because we previously check for the latest snapshot. But, if it
-        // happens, instead of crashing raft thread, we respond with success because we already have
-        // the snapshot.
+            raft.getPersistedSnapshotStore()
+                .newReceivedSnapshot(snapshotChunk.getSnapshotId())
+                .get();
+      } catch (final ExecutionException errorCreatingPendingSnapshot) {
+        return failIfSnapshotAlreadyExists(errorCreatingPendingSnapshot, snapshotChunk);
+      } catch (final InterruptedException e) {
+        log.warn(
+            "Failed to create pending snapshot when receiving snapshot {}",
+            snapshotChunk.getSnapshotId(),
+            e);
         return CompletableFuture.completedFuture(
-            logResponse(InstallResponse.builder().withStatus(RaftResponse.Status.OK).build()));
+            logResponse(
+                InstallResponse.builder()
+                    .withStatus(Status.ERROR)
+                    .withError(Type.APPLICATION_ERROR, "Failed to create pending snapshot")
+                    .build()));
       }
+
       log.info("Started receiving new snapshot {} from {}", pendingSnapshot, request.leader());
       pendingSnapshotStartTimestamp = System.currentTimeMillis();
       snapshotReplicationMetrics.incrementCount();
@@ -378,6 +389,29 @@ public class PassiveRole extends InactiveRole {
                 .withError(
                     RaftError.Type.ILLEGAL_MEMBER_STATE, "Cannot request vote from RESERVE member")
                 .build()));
+  }
+
+  private CompletableFuture<InstallResponse> failIfSnapshotAlreadyExists(
+      final ExecutionException errorCreatingPendingSnapshot,
+      final SnapshotChunkImpl snapshotChunk) {
+    if (errorCreatingPendingSnapshot.getCause() instanceof SnapshotAlreadyExistsException) {
+      // This should not happen because we previously check for the latest snapshot. But, if it
+      // happens, instead of crashing raft thread, we respond with success because we already
+      // have the snapshot.
+      return CompletableFuture.completedFuture(
+          logResponse(InstallResponse.builder().withStatus(Status.OK).build()));
+    } else {
+      log.warn(
+          "Failed to create pending snapshot when receiving snapshot {}",
+          snapshotChunk.getSnapshotId(),
+          errorCreatingPendingSnapshot);
+      return CompletableFuture.completedFuture(
+          logResponse(
+              InstallResponse.builder()
+                  .withStatus(Status.ERROR)
+                  .withError(Type.APPLICATION_ERROR, "Failed to create pending snapshot")
+                  .build()));
+    }
   }
 
   private void onSnapshotReceiveCompletedOrAborted() {

--- a/atomix/cluster/src/test/java/io/atomix/cluster/NoopSnapshotStore.java
+++ b/atomix/cluster/src/test/java/io/atomix/cluster/NoopSnapshotStore.java
@@ -78,7 +78,7 @@ public class NoopSnapshotStore implements ReceivableSnapshotStore {
   }
 
   @Override
-  public ReceivedSnapshot newReceivedSnapshot(final String snapshotId) {
+  public ActorFuture<ReceivedSnapshot> newReceivedSnapshot(final String snapshotId) {
     return null;
   }
 

--- a/atomix/cluster/src/test/java/io/atomix/raft/snapshot/TestSnapshotStore.java
+++ b/atomix/cluster/src/test/java/io/atomix/raft/snapshot/TestSnapshotStore.java
@@ -114,17 +114,19 @@ public class TestSnapshotStore implements ReceivableSnapshotStore {
   }
 
   @Override
-  public ReceivedSnapshot newReceivedSnapshot(final String snapshotId) {
+  public ActorFuture<ReceivedSnapshot> newReceivedSnapshot(final String snapshotId) {
     if (Optional.ofNullable(currentPersistedSnapshot.get())
         .map(PersistedSnapshot::getId)
         .orElse("")
         .equals(snapshotId)) {
-      throw new SnapshotAlreadyExistsException("Snapshot with this ID is already persisted");
+      CompletableActorFuture.completedExceptionally(
+          new SnapshotAlreadyExistsException("Snapshot with this ID is already persisted"));
     }
 
     final var newSnapshot = new InMemorySnapshot(this, snapshotId);
     receivedSnapshots.add(newSnapshot);
-    return newSnapshot;
+
+    return CompletableActorFuture.completed(newSnapshot);
   }
 
   @Override

--- a/snapshot/src/main/java/io/camunda/zeebe/snapshots/ReceivableSnapshotStore.java
+++ b/snapshot/src/main/java/io/camunda/zeebe/snapshots/ReceivableSnapshotStore.java
@@ -7,6 +7,8 @@
  */
 package io.camunda.zeebe.snapshots;
 
+import io.camunda.zeebe.scheduler.future.ActorFuture;
+
 public interface ReceivableSnapshotStore extends PersistedSnapshotStore {
 
   /**
@@ -16,5 +18,5 @@ public interface ReceivableSnapshotStore extends PersistedSnapshotStore {
    *     index-term-timestamp-processedposition-exportedposition}
    * @return the new volatile received snapshot
    */
-  ReceivedSnapshot newReceivedSnapshot(String snapshotId);
+  ActorFuture<? extends ReceivedSnapshot> newReceivedSnapshot(String snapshotId);
 }

--- a/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java
+++ b/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java
@@ -30,11 +30,8 @@ import io.camunda.zeebe.util.Either;
 import io.camunda.zeebe.util.FileUtil;
 import java.io.IOException;
 import java.io.UncheckedIOException;
-import java.nio.file.AtomicMoveNotSupportedException;
-import java.nio.file.FileAlreadyExistsException;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.nio.file.StandardCopyOption;
 import java.util.Collections;
 import java.util.ConcurrentModificationException;
 import java.util.HashSet;
@@ -44,7 +41,6 @@ import java.util.Set;
 import java.util.concurrent.CopyOnWriteArraySet;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.stream.Collectors;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -291,10 +287,7 @@ public final class FileBasedSnapshotStore extends Actor
     final CompletableActorFuture<Void> abortFuture = new CompletableActorFuture<>();
     actor.run(
         () -> {
-          final var abortedAll =
-              pendingSnapshots.stream()
-                  .map(PersistableSnapshot::abort)
-                  .collect(Collectors.toList());
+          final var abortedAll = pendingSnapshots.stream().map(PersistableSnapshot::abort).toList();
           actor.runOnCompletion(
               abortedAll,
               error -> {
@@ -594,21 +587,6 @@ public final class FileBasedSnapshotStore extends Actor
     purgePendingSnapshots(newPersistedSnapshot.getSnapshotId());
   }
 
-  private void moveToSnapshotDirectory(final Path directory, final Path destination) {
-    try {
-      tryAtomicDirectoryMove(directory, destination);
-    } catch (final FileAlreadyExistsException e) {
-      LOGGER.debug(
-          "Expected to move snapshot from {} to {}, but it already exists",
-          directory,
-          destination,
-          e);
-    } catch (final IOException e) {
-      rollbackPartialSnapshot(destination);
-      throw new UncheckedIOException(e);
-    }
-  }
-
   private void rollbackPartialSnapshot(final Path destination) {
     try {
       FileUtil.deleteFolderIfExists(destination);
@@ -628,20 +606,6 @@ public final class FileBasedSnapshotStore extends Actor
     } catch (final IOException e) {
       LOGGER.warn("Failed to delete not completed (orphaned) snapshot {}", pendingSnapshot, e);
     }
-  }
-
-  private void tryAtomicDirectoryMove(final Path directory, final Path destination)
-      throws IOException {
-    try {
-      FileUtil.moveDurably(directory, destination, StandardCopyOption.ATOMIC_MOVE);
-    } catch (final AtomicMoveNotSupportedException e) {
-      LOGGER.warn("Atomic move not supported. Moving the snapshot files non-atomically", e);
-      FileUtil.moveDurably(directory, destination);
-    }
-  }
-
-  private Path buildPendingSnapshotDirectory(final SnapshotId id) {
-    return pendingDirectory.resolve(id.getSnapshotIdAsString());
   }
 
   private Path buildSnapshotDirectory(final FileBasedSnapshotId snapshotId) {

--- a/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java
+++ b/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java
@@ -355,20 +355,16 @@ public final class FileBasedSnapshotStore extends Actor
                         + snapshotId
                         + "'."));
 
-    final var directory = buildSnapshotDirectory(parsedSnapshotId);
-    if (directory.toFile().exists()) {
-      actor.run(
-          () -> {
-            try {
-              checkAndCleanupExistingDirectory(snapshotId, parsedSnapshotId, directory);
-              createReceivedSnapshot(parsedSnapshotId, directory, newSnapshotFuture);
-            } catch (final Exception e) {
-              newSnapshotFuture.completeExceptionally(e);
-            }
-          });
-    } else {
-      createReceivedSnapshot(parsedSnapshotId, directory, newSnapshotFuture);
-    }
+    actor.run(
+        () -> {
+          final var directory = buildSnapshotDirectory(parsedSnapshotId);
+          try {
+            checkAndCleanupExistingDirectory(snapshotId, parsedSnapshotId, directory);
+            createReceivedSnapshot(parsedSnapshotId, directory, newSnapshotFuture);
+          } catch (final Exception e) {
+            newSnapshotFuture.completeExceptionally(e);
+          }
+        });
     return newSnapshotFuture;
   }
 

--- a/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java
+++ b/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java
@@ -344,7 +344,8 @@ public final class FileBasedSnapshotStore extends Actor
   }
 
   @Override
-  public FileBasedReceivedSnapshot newReceivedSnapshot(final String snapshotId) {
+  public ActorFuture<FileBasedReceivedSnapshot> newReceivedSnapshot(final String snapshotId) {
+    final var newSnapshotFuture = new CompletableActorFuture<FileBasedReceivedSnapshot>();
     final var optSnapshotId = FileBasedSnapshotId.ofFileName(snapshotId);
     final var parsedSnapshotId =
         optSnapshotId.orElseThrow(
@@ -355,6 +356,34 @@ public final class FileBasedSnapshotStore extends Actor
                         + "'."));
 
     final var directory = buildSnapshotDirectory(parsedSnapshotId);
+    if (directory.toFile().exists()) {
+      actor.run(
+          () -> {
+            try {
+              checkAndCleanupExistingDirectory(snapshotId, parsedSnapshotId, directory);
+              createReceivedSnapshot(parsedSnapshotId, directory, newSnapshotFuture);
+            } catch (final Exception e) {
+              newSnapshotFuture.completeExceptionally(e);
+            }
+          });
+    } else {
+      createReceivedSnapshot(parsedSnapshotId, directory, newSnapshotFuture);
+    }
+    return newSnapshotFuture;
+  }
+
+  private void createReceivedSnapshot(
+      final FileBasedSnapshotId parsedSnapshotId,
+      final Path directory,
+      final CompletableActorFuture<FileBasedReceivedSnapshot> newSnapshotFuture) {
+    final var newPendingSnapshot =
+        new FileBasedReceivedSnapshot(parsedSnapshotId, directory, this, actor);
+    addPendingSnapshot(newPendingSnapshot);
+    newSnapshotFuture.complete(newPendingSnapshot);
+  }
+
+  private void checkAndCleanupExistingDirectory(
+      final String snapshotId, final FileBasedSnapshotId parsedSnapshotId, final Path directory) {
     if (directory.toFile().exists()) {
       if (!buildSnapshotsChecksumPath(parsedSnapshotId).toFile().exists()) {
         try {
@@ -374,10 +403,6 @@ public final class FileBasedSnapshotStore extends Actor
                 snapshotId));
       }
     }
-    final var newPendingSnapshot =
-        new FileBasedReceivedSnapshot(parsedSnapshotId, directory, this, actor);
-    addPendingSnapshot(newPendingSnapshot);
-    return newPendingSnapshot;
   }
 
   @Override

--- a/snapshot/src/test/java/io/camunda/zeebe/snapshots/PersistedSnapshotStoreTest.java
+++ b/snapshot/src/test/java/io/camunda/zeebe/snapshots/PersistedSnapshotStoreTest.java
@@ -70,7 +70,7 @@ public class PersistedSnapshotStoreTest {
     final var index = 1L;
 
     // when
-    final var transientSnapshot = persistedSnapshotStore.newReceivedSnapshot("1-0-123-121");
+    final var transientSnapshot = persistedSnapshotStore.newReceivedSnapshot("1-0-123-121").join();
 
     // then
     assertThat(transientSnapshot.index()).isEqualTo(index);

--- a/snapshot/src/test/java/io/camunda/zeebe/snapshots/PersistedSnapshotStoreTest.java
+++ b/snapshot/src/test/java/io/camunda/zeebe/snapshots/PersistedSnapshotStoreTest.java
@@ -28,7 +28,9 @@ public class PersistedSnapshotStoreTest {
     final var partitionId = 1;
     final var root = temporaryFolder.getRoot();
 
-    persistedSnapshotStore = new FileBasedSnapshotStore(partitionId, root.toPath());
+    final var snapshotStore = new FileBasedSnapshotStore(partitionId, root.toPath());
+    scheduler.submitActor(snapshotStore).join();
+    persistedSnapshotStore = snapshotStore;
   }
 
   @Test

--- a/snapshot/src/test/java/io/camunda/zeebe/snapshots/ReceivedSnapshotTest.java
+++ b/snapshot/src/test/java/io/camunda/zeebe/snapshots/ReceivedSnapshotTest.java
@@ -25,7 +25,9 @@ import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardOpenOption;
+import java.time.Duration;
 import java.util.Map;
+import java.util.concurrent.ExecutionException;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
@@ -63,7 +65,7 @@ public class ReceivedSnapshotTest {
     // given
 
     // when
-    assertThatThrownBy(() -> receiverSnapshotStore.newReceivedSnapshot("invalid"))
+    assertThatThrownBy(() -> receiverSnapshotStore.newReceivedSnapshot("invalid").join())
         .isInstanceOf(IllegalArgumentException.class);
   }
 
@@ -72,7 +74,7 @@ public class ReceivedSnapshotTest {
     // given
 
     // when
-    final var receivedSnapshot = receiverSnapshotStore.newReceivedSnapshot("1-0-123-121");
+    final var receivedSnapshot = receiverSnapshotStore.newReceivedSnapshot("1-0-123-121").join();
 
     // then
     assertThat(receivedSnapshot.index()).isEqualTo(1L);
@@ -154,7 +156,7 @@ public class ReceivedSnapshotTest {
 
     // when
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
 
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       final SnapshotChunk chunk = snapshotChunkReader.next();
@@ -182,9 +184,11 @@ public class ReceivedSnapshotTest {
     final var firstPersistedSnapshot = firstReceivedSnapshot.persist().join();
 
     // when then throw - receives same snapshot again
-    assertThatThrownBy(() -> receiveSnapshot(persistedSnapshot))
-        .isInstanceOf(SnapshotAlreadyExistsException.class)
-        .hasMessageContaining(
+    assertThat(receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()))
+        .failsWithin(Duration.ofMillis(100))
+        .withThrowableOfType(ExecutionException.class)
+        .withCauseInstanceOf(SnapshotAlreadyExistsException.class)
+        .withMessageContaining(
             "Expected to receive snapshot with id 1-0-1-0, but was already persisted");
   }
 
@@ -193,7 +197,7 @@ public class ReceivedSnapshotTest {
     // given
     final var persistedSnapshot = takePersistedSnapshot();
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
 
     // when
     assertThatThrownBy(() -> receivedSnapshot.persist().join())
@@ -205,7 +209,7 @@ public class ReceivedSnapshotTest {
     // given
     final var persistedSnapshot = takePersistedSnapshot();
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       receivedSnapshot.apply(snapshotChunkReader.next()).join();
     }
@@ -227,7 +231,7 @@ public class ReceivedSnapshotTest {
     // given
     final var persistedSnapshot = takePersistedSnapshot();
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
 
     // when - then
     assertThatCode(receivedSnapshot::abort).doesNotThrowAnyException();
@@ -286,7 +290,7 @@ public class ReceivedSnapshotTest {
 
     // when
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
 
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       receivedSnapshot.apply(snapshotChunkReader.next()).join();
@@ -309,7 +313,7 @@ public class ReceivedSnapshotTest {
 
     // when
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       final SnapshotChunk originalChunk = snapshotChunkReader.next();
       final SnapshotChunk corruptedChunk =
@@ -329,7 +333,7 @@ public class ReceivedSnapshotTest {
 
     // when
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       while (snapshotChunkReader.hasNext()) {
         final var originalChunk = snapshotChunkReader.next();
@@ -353,7 +357,7 @@ public class ReceivedSnapshotTest {
 
     // when
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       final var firstChunk = snapshotChunkReader.next();
       receivedSnapshot.apply(firstChunk).join();
@@ -372,7 +376,7 @@ public class ReceivedSnapshotTest {
 
     // when
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       final var firstChunk = snapshotChunkReader.next();
       receivedSnapshot.apply(firstChunk).join();
@@ -394,7 +398,7 @@ public class ReceivedSnapshotTest {
 
     // when
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       while (snapshotChunkReader.hasNext()) {
         final var corruptedChunk =
@@ -419,7 +423,7 @@ public class ReceivedSnapshotTest {
 
     // when
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       final var originalChunk = snapshotChunkReader.next();
       final SnapshotChunk corruptedChunk = SnapshotChunkWrapper.withSnapshotId(originalChunk, "id");
@@ -433,7 +437,7 @@ public class ReceivedSnapshotTest {
 
   private ReceivedSnapshot receiveSnapshot(final PersistedSnapshot persistedSnapshot) {
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
 
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       while (snapshotChunkReader.hasNext()) {

--- a/snapshot/src/test/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshotTest.java
+++ b/snapshot/src/test/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshotTest.java
@@ -69,7 +69,7 @@ public class FileBasedReceivedSnapshotTest {
 
     // when
     final ReceivedSnapshot receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot("1-0-123-121");
+        receiverSnapshotStore.newReceivedSnapshot("1-0-123-121").join();
 
     // then
     assertThat(receivedSnapshot.getPath())
@@ -97,7 +97,7 @@ public class FileBasedReceivedSnapshotTest {
     // given
     final var persistedSnapshot = takePersistedSnapshot(1L);
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
 
     // when
     final SnapshotChunk expectedChunk;
@@ -224,7 +224,7 @@ public class FileBasedReceivedSnapshotTest {
 
     // when
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
 
     final SnapshotChunk firstChunk;
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
@@ -251,7 +251,7 @@ public class FileBasedReceivedSnapshotTest {
     // when
     final SnapshotChunk firstChunk;
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       firstChunk = snapshotChunkReader.next();
       receivedSnapshot.apply(firstChunk).join();
@@ -276,7 +276,7 @@ public class FileBasedReceivedSnapshotTest {
     // when
     final SnapshotChunk firstChunk;
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       firstChunk = snapshotChunkReader.next();
       receivedSnapshot.apply(firstChunk).join();
@@ -303,7 +303,7 @@ public class FileBasedReceivedSnapshotTest {
     // when
     final SnapshotChunk firstChunk;
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       firstChunk = snapshotChunkReader.next();
       receivedSnapshot.apply(firstChunk).join();
@@ -342,7 +342,7 @@ public class FileBasedReceivedSnapshotTest {
 
   private ReceivedSnapshot receiveSnapshot(final PersistedSnapshot persistedSnapshot) {
     final var receivedSnapshot =
-        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId());
+        receiverSnapshotStore.newReceivedSnapshot(persistedSnapshot.getId()).join();
 
     try (final var snapshotChunkReader = persistedSnapshot.newChunkReader()) {
       while (snapshotChunkReader.hasNext()) {

